Fix part of #5485: Create means for verifying Fragment Arguments

subhajitxyz opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation

Fixes part of #5485

Added tests for 11 fragment arguments and 'saveInstanceState'.

I have not added tests for the following:
1. `onSaveInstanceState` of `StateFragment`.
2. `OptionsFragment`.

Problems:
1. When trying to recreate `StateFragment`, an error occurs: "State has not been initialized."
2. In the `OptionsFragmentTest` class, when I try to run the test on Robolectric, it shows "No tests were found."


## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

subhajitxyz committed:
Merge pull request #1 from oppia/develop

Develop (fork) Branch Update

subhajitxyz committed:
Merge branch 'oppia:develop' into develop

subhajitxyz committed:
Merge remote-tracking branch 'upstream/develop' into develop

subhajitxyz committed:
Merge pull request #2 from oppia/develop

Develop (fork) Branch Update #2

subhajitxyz committed:
Merge pull request #3 from oppia/develop

Develop (fork) Branch Update

subhajitxyz committed:
Merge pull request #4 from oppia/develop

Develop(fork) branch update

subhajitxyz committed:
Merge remote-tracking branch 'upstream/develop' into develop

subhajitxyz committed:
Merge pull request #5 from oppia/develop

Develop (fork) Branch Update

subhajitxyz committed:
Tests are added for fragment arguments and onsaveinstance

subhajitxyz committed:
Corrected TopicInfoFragmentTest

subhajitxyz committed:
Solved code formatting

subhajitxyz committed:
Solved Klint issue

subhajitxyz committed:
Correct Kdoc formatting

subhajitxyz commented:
@adhiamboperes ,PTAL.

theMr17 commented:
Hi @subhajitxyz, Will there be more PRs for this issue? If so, please update the PR title to "Fix part of #5485" and adjust the description accordingly.

Your previous PRs had a similar issue with the title, which caused the issue to close automatically and had to be reopened manually.

subhajitxyz commented:
Thanks @theMr17 for addressing the problem. I will be more careful in the future. There are no more PRs for this issue.

theMr17 commented:
@subhajitxyz, I just wanted to confirm if there are any remaining fragments, aside from `StateFragment` and `OptionsFragment`, that still need to be addressed after this PR is merged?

subhajitxyz commented:
There won't be any remaining fragments left to address aside from the StateFragment and OptionsFragment.

theMr17 commented on app/src/sharedTest/java/org/oppia/android/app/policies/PoliciesFragmentTest.kt:
```diff
@@ -325,6 +329,44 @@ class PoliciesFragmentTest {
+        }
+        val receivedPolicies =
+          args.getProto(
+            "PoliciesFragment.policy_page",
```
```suggestion
            PoliciesFragmentTestActivity.POLICIES_FRAGMENT_TEST_POLICY_PAGE_PARAMS_PROTO,
```

subhajitxyz replied:
Hi @theMr17 ,I think , we should use  PoliciesFragment.POLICIES_FRAGMENT_POLICY_PAGE_ARGUMENT_PROTO.
to use it , we need to set POLICIES_FRAGMENT_POLICY_PAGE_ARGUMENT_PROTO as "Public" in PoliciesFragment.


If we use PoliciesFragmentTestActivity.POLICIES_FRAGMENT_TEST_POLICY_PAGE_PARAMS_PROTO. we are getting argument of PoliciesFragmentTestActivity and the test is failed.

Could you please review it once and confirm what should i use?

theMr17 replied:
You are right, it should be `PoliciesFragment.POLICIES_FRAGMENT_POLICY_PAGE_ARGUMENT_PROTO`. Please update it accordingly.

Making it public should be fine.

subhajitxyz replied:
Okay, fine. I'll make the update accordingly.

subhajitxyz replied:
Done.

theMr17 commented on app/src/sharedTest/java/org/oppia/android/app/thirdparty/LicenseTextViewerFragmentTest.kt:
```diff
@@ -317,6 +321,36 @@ class LicenseTextViewerFragmentTest {
+          "Expected arguments to be passed to LicenseTextViewerFragment"
+        }
+        val args = arguments.getProto(
+          "LicenseTextViewerFragment.arguments",
```
```suggestion
          LicenseTextViewerFragment.LICENSE_TEXT_VIEWER_FRAGMENT_ARGUMENTS_KEY,
```

subhajitxyz replied:
We need to make LICENSE_TEXT_VIEWER_FRAGMENT_ARGUMENTS_KEY public  in LicenseTextViewerFragment to use it here. 
Could you please confirm if this change is acceptable?

theMr17 replied:
Making it public should be fine.

theMr17 commented on app/src/sharedTest/java/org/oppia/android/app/topic/revision/TopicRevisionFragmentTest.kt:
```diff
@@ -286,6 +292,40 @@ class TopicRevisionFragmentTest {
   }
 
+  @Test
+  fun testFragment_fragmentLoaded_verifyCorrectArgumentsPassed() {
```
```suggestion
  fun testTopicRevisionFragment_fragmentLoaded_verifyCorrectArgumentsPassed() {
```

subhajitxyz replied:
Done.

theMr17 commented on app/src/sharedTest/java/org/oppia/android/app/walkthrough/WalkthroughFinalFragmentTest.kt:
```diff
@@ -268,6 +273,33 @@ class WalkthroughFinalFragmentTest {
   }
 
+  @Test
+  fun testFragment_fragmentLoaded_verifyCorrectArgumentsPassed() {
```
```suggestion
  fun testWalkthroughFinalFragment_fragmentLoaded_verifyCorrectArgumentsPassed() {
```

subhajitxyz replied:
Done.

theMr17 commented on app/src/sharedTest/java/org/oppia/android/app/topic/lessons/TopicLessonsFragmentTest.kt:
```diff
@@ -1205,6 +1211,92 @@ class TopicLessonsFragmentTest {
   }
 
+  @Test
+  fun testFragment_fragmentLoaded_verifyCorrectArgumentsPassed() {
```
```suggestion
  fun testLessonPlayFragment_fragmentLoaded_verifyCorrectArgumentsPassed() {
```
Ditto for all other tests.

subhajitxyz replied:
Done.

theMr17 reviewed (changes requested):
Thanks, @subhajitxyz! I've left some comments, PTAL.

I'm updating the PR to partially resolve the issue, as it needs to remain open for `OptionsFragment` and `StateFragment`, which will be addressed at a later time.

subhajitxyz committed:
Changes as per review suggestion

subhajitxyz committed:
Merge remote-tracking branch 'upstream/develop' into fragment-tests-part3

subhajitxyz committed:
Made changes in PoliciesFragmentTest

subhajitxyz committed:
Fix Kdoc in PoliciesFragment

subhajitxyz commented:
Hi @theMr17 ,PTAL.

theMr17 reviewed (approved):
Thanks @subhajitxyz, this LGTM!

theMr17 commented:
@BenHenning PTAL.

BenHenning commented on app/src/main/java/org/oppia/android/app/topic/lessons/TopicLessonsFragment.kt:
```diff
@@ -145,4 +145,14 @@ class TopicLessonsFragment :
     topicLessonsFragmentPresenter.selectChapterSummary(storyId, explorationId, chapterPlayState)
   }
+
+  /** Returns the [currentExpandedChapterListIndex] stored in the fragment. */
+  fun getCurrentExpandedChapterListIndex(): Int? {
+    return currentExpandedChapterListIndex
+  }
+
+  /** Returns the [isDefaultStoryExpanded] stored in the fragment. */
+  fun getIsDefaultStoryExpanded(): Boolean {
+    return isDefaultStoryExpanded
+  }
```
Rather than exposing internal state variables for test verification, is it possible to verify that the UI is in the correct state such that we can infer these? For example, could we check that the (expected) default story is currently expanded?

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/administratorcontrols/AdministratorControlsFragmentTest.kt:
```diff
@@ -434,6 +438,36 @@ class AdministratorControlsFragmentTest {
+        )
+        val receivedIsMultipane = args.isMultipane
+
+        assertThat(receivedIsMultipane).isEqualTo(isMultipane)
```
Will 'isMultipane' always be true/false based on the configuration of the test? It would probably be more robust to set up the test such that we can always verify true/false as expected per the test setup (and add a second test for the opposite case so that both 'true' and 'false' are explicitly verified).

subhajitxyz replied:
@BenHenning , Please check, Could we do like this way?
Create variable isMultipane in AdministratorControlsFragmentTestActivityPresenter. 
while testing for ismultipane true, i am setting isMultipane true and calling presenter.handleOnCreate(). 

    /** The presenter for [AdministratorControlsFragmentTestActivity]. */
    @ActivityScope
    class AdministratorControlsFragmentTestActivityPresenter @Inject constructor(
      private val activity: AppCompatActivity
    ) {
      var isMultipane = false
    
      /** Initializes the [AdministratorControlsFragmentTestActivity] and sets the navigation drawer. */
      fun handleOnCreate() {
        activity.setContentView(R.layout.administrator_controls_fragment_test_activity)
        activity.supportFragmentManager.beginTransaction().replace(
          R.id.administrator_controls_fragment_test_activity_fragment_container,
          AdministratorControlsFragment.newInstance(isMultipane)
        ).commitNow()
      }
    }
    
    ///testing when is multipane is set true.
    
    @Test
      fun testFragmentArguments_afterCreationWithIsMultipaneTrue_areCorrect() {
        launch<AdministratorControlsFragmentTestActivity>(
          createAdministratorControlsFragmentTestActivityIntent(
            profileId = internalProfileId
          )
        ).use { scenario ->
          testCoroutineDispatchers.runCurrent()
          scenario.onActivity { activity ->
            val presenter = activity.administratorControlsFragmentTestActivityPresenter
            presenter.isMultipane = true // Set `isMultipane` to true
            presenter.handleOnCreate()
    
            val administratorControlsFragment = activity.supportFragmentManager
              .findFragmentById(R.id.administrator_controls_fragment_test_activity_fragment_container)
              as AdministratorControlsFragment
    
            val arguments = checkNotNull(administratorControlsFragment.arguments) {
              "Expected arguments to be passed to AdministratorControlsFragment"
            }
            val args = arguments.getProto(
              ADMINISTRATOR_CONTROLS_FRAGMENT_ARGUMENTS_KEY,
              AdministratorControlsFragmentArguments.getDefaultInstance()
            )
            val receivedIsMultipane = args.isMultipane
    
            assertThat(receivedIsMultipane).isEqualTo(true)
          }
        }
      }

BenHenning replied:
I don't think that will work--the activity lifecycle methods need to be handled by Android (and specifically Robolectric when running the tests locally).

Normally multi-pane is automatically determined in production code by checking whether the current device is a tablet. We're able to configure tests to run with a tablet display environment (using per-test qualifiers). Would that work in this case? Here's one example of such a test configuration:

https://github.com/oppia/oppia-android/blob/55ab73edfbd032045a065986ac06db7d2778bf48/app/src/sharedTest/java/org/oppia/android/app/help/HelpFragmentTest.kt#L324

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt:
```diff
@@ -5004,6 +5006,39 @@ class StateFragmentTest {
   }
 
+  @Test
+  fun testStateFragment_fragmentLoaded_verifyCorrectArgumentsPassed() {
```
```suggestion
  fun testFragmentArguments_afterCreation_areCorrect() {
```

Slightly more deliberate way to name what's being tested. Ditto for all other tests in this PR.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt:
```diff
@@ -5004,6 +5006,39 @@ class StateFragmentTest {
+      FRACTIONS_EXPLORATION_ID_1,
+      shouldSavePartialProgress = false
+    ).use { scenario ->
+      testCoroutineDispatchers.unregisterIdlingResource()
```
Why unregister the idling resource? This seems rather unusual.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/policies/PoliciesFragmentTest.kt:
```diff
@@ -325,6 +329,44 @@ class PoliciesFragmentTest {
+            PoliciesFragmentArguments.getDefaultInstance()
+          )
+
+        assertThat(receivedPolicies).isEqualTo(policiesFragmentArguments)
```
Prefer checking the properties of the proto rather than the whole thing (since it helps make the test more deliberate and contextual).

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/topic/revision/TopicRevisionFragmentTest.kt:
```diff
@@ -286,6 +292,40 @@ class TopicRevisionFragmentTest {
+          TopicRevisionFragment.TOPIC_REVISION_FRAGMENT_ARGUMENTS_KEY,
+          TopicRevisionFragmentArguments.getDefaultInstance()
+        )
+        val receivedTopicId = checkNotNull(args?.topicId) {
+          "Expected topic ID to be included in arguments for TopicRevisionFragment."
+        }
```
FYI in most of the new test code you can remove the extra check*() logic and just have Truth do the work. E.g. in this case:

```kotlin
assertThat(args?.topicId).isEqualTo(FRACTIONS_TOPIC_ID)
```

Since Truth is a Java API, it can take null values (and will fail correctly if it receives one since, in this case, ``FRACTIONS_TOPIC_ID`` is not null and will trigger a failure if ``args`` is null).

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/topic/practice/TopicPracticeFragmentTest.kt:
```diff
@@ -383,6 +389,91 @@ class TopicPracticeFragmentTest {
+        val restoredSkillIdHashMap = topicPracticeFragment.topicPracticeFragmentPresenter
+          .skillIdHashMap
+
+        assertThat(restoredSelectedSubtopicList).containsExactly(1, 2, 3)
+        assertThat(restoredSkillIdHashMap)
+          .containsExactly(
+            1, mutableListOf("skill_1"),
+            2, mutableListOf("skill_2", "skill_3")
+          )
```
Here & above: similar to my earlier comment, can we test this by analyzing the UI rather than inspecting fragment state? The latter is rather fragile if the implementation state changes, and technically isn't representative of the thing actually being represented by the values.

BenHenning committed:
Merge branch 'develop' into fragment-tests-part3

BenHenning reviewed (changes requested):
Thanks @subhajitxyz! Super sorry for the delayed review.

Had a few comments--PTAL.

adhiamboperes committed:
Fix Part of #4938: Language Selection Config and New Profile Creation Flow (#5457)

<!-- READ ME FIRST: Please fill in the explanation section below and
check off every point from the Essential Checklist! -->
## Explanation
Fixes Part of #4938: Modifies the profile creation flow and sets the
app/audio language selection during onboarding.

### Default Profile Creation
A default empty profile is created when the Onboarding screen is opened
for the first time. This is necessary to provide a ProfileId to use when
saving the selected app language.

Because this will be the first profile on the app in onboarding v2, it
is an admin.

Provisions have been made so that, should the user exit the app before
completing the onboarding flow, this profile will be fetched, to prevent
multiple profile creation.




### App Language Selection
The language selector will be shown to the user on initial app launch,
or if profile onboarding is not complete.

- There will be a pre-filled language option based on the locale of the
device when the app is installed. If the locale is unsupported, English
will be the default selection.
- A user can select any preferred supported app language from the
dropdown list.
- The existing language functionality/behavior will be retained.

Tests have been added to verify these requirements, and efforts have
been made to ensure language selection persists on configuration change.
I noticed during testing that failure to do this resulted in an
unpleasant UX.

### Profile Nickname and Picture
The "Create profile Screen" has been repurposed to update the default
profile instead, providing the remaining profile properties(ProfileType,
name, avatar)

Checks have been added to check for profile creation errors, consistent
with the legacy flow.

A new function has been added to the ProfileManagementController to
allow for batch update of these fields, and corresponding tests have
been added.

### Audio Language Selection
The final step of onboarding is the audio language selection, and there
will be a pre-filled language selectionas follows:

- Selected app language(from first onboarding screen) if available as
audio language. (Audio language need not be completed)
- Otherwise (if available as audio language), audio language of the
administrator account -- this will be added downstream in M3, as it only
impacts additional learners.
- Otherwise (if available as audio language), device language.
- Else, English.

There have been some incidental changes in AudioLanguageFragment and
OptionsFragment, and their related tests due to sharing of the classes
between the existing screens and the new screens.

### ProfileTestHelper.kt
A new function, and related tests, have been added to create a default
profile to be used in tests.

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If
this PR fixes part of an issue, prefix the title with "Fix part of
#bugnum: ...".)
- [x] Any changes to
[scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets)
files have their rationale included in the PR explanation.
- [x] The PR follows the [style
guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android
Studio
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and
is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
The screen recordings are in [this drive
location](https://drive.google.com/drive/folders/1CXTAALPgpCKfekOQKjRFTb5G2fPgWkRt?usp=sharing),
since github does not support webm.

---------

Co-authored-by: Ben Henning <ben@oppia.org>

kkmurerwa committed:
Fixes multiple onboarding events issue (#5551)

<!-- READ ME FIRST: Please fill in the explanation section below and
check off every point from the Essential Checklist! -->
## Explanation
<!--
- Explain what your PR does. If this PR fixes an existing bug, please
include
- "Fixes #bugnum:" in the explanation so that GitHub can auto-close the
issue
  - when this PR is merged.
  -->
Fixes multiple onboarding events issue:

From the previous implementation of the onboarding logs, it was possible
to have multiple onboarding logs belonging to the same user. This
happened because the onboarding flag used when making the log was not
updated until an app restart. For this reason, if a user went back to
the home screen several times or if the user changed orientation from
portrait to landscape and vice-versa, the onboarding complete log would
be triggered every time this happened.

This PR intends to fix this issue by centralizing the onboarding
complete log. It will also allow an optional profileId variable to be
passed as support for Onboarding V2 where onboarding will be tied to an
individual user.

## Manual testing process and results
To confirm that the bug was fixed, I manually performed a test using the
steps outlined below;

1. Clean the Oppia app storage from settings or uninstall the Oppia app
to prime it for the test then reinstall it.
2. Go through the onboarding screens, then select Administrator when you
reach the profile selection screen.
3. This opens the home screen. Click the hamburger menu at the top right
corner of the screen.
4. Click **Developer options** at the bottom of the side-bar menu. 
5. Under the **Event logs** section, click the event logs item. You
should see a list of events logged so far, including a **Complete App
Onboarding** event right above the **Open Home** event.
7. Click the back arrow at the top right corner to go back to the
developer options menu and then click the hamburger menu to open the
side pane again.
8. Click on Home to go back to the home screen and then repeat steps 3
to 8, three more times.
9. Once that is done, reopen the Event Logs screen. With the original
buggy flow, three Open Home and Complete App Onboarding events should be
displayed. With this fix, only one Complete App Onboarding event should
be logged but 3 Open Home events should be displayed.

|Buggy flow|New flow with fix|
|----|----|

|![Screenshot_20241021_102501](https://github.com/user-attachments/assets/8c6aaf59-4e10-4382-895d-619ffc2b1f18)|![Screenshot_20241021_101850](https://github.com/user-attachments/assets/ca576bb2-33e2-4ff5-ae19-8e7b0444474c)|

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If
this PR fixes part of an issue, prefix the title with "Fix part of
#bugnum: ...".)
- [x] Any changes to
[scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets)
files have their rationale included in the PR explanation.
- [x] The PR follows the [style
guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android
Studio
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and
is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related
changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of
the before & after UI changes
- For the screenshots above, include both English and pseudo-localized
(RTL) screenshots (see [RTL
guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see
[accessibility
guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and
dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests
locally & that they're passing

---------

Co-authored-by: Kenneth Murerwa <kenneth.murerwa@kopokopo.com>

Mayank77maruti committed:
Fix #5547 : Added Developer Onboarding Notification (#5549)

<!-- READ ME FIRST: Please fill in the explanation section below and
check off every point from the Essential Checklist! -->
Fixes #5547 

This PR introduces a new GitHub Action that automatically posts a
comment on a pull request once it has been merged into the develop
branch. The comment provides different feedback based on whether it is
the user's first or second merged PR to the repository.

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If
this PR fixes part of an issue, prefix the title with "Fix part of
#bugnum: ...".)
- [x] Any changes to
[scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets)
files have their rationale included in the PR explanation.
- [x] The PR follows the [style
guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android
Studio
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and
is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

subhajitxyz commented on app/src/sharedTest/java/org/oppia/android/app/topic/lessons/TopicLessonsFragmentTest.kt:
```diff
@@ -1205,6 +1211,92 @@ class TopicLessonsFragmentTest {
+        assertThat(receivedCurrentExpandedChapterListIndex).isEqualTo(1)
+        assertThat(receivedIsDefaultStoryExpanded).isEqualTo(true)
+      }
+    }
```
@BenHenning ,Could we test like that without exposing internal variable of TopicLessonsFragment (currentExpandedChapterListIndex,  isDefaultStoryExpanded) ? As you mention above.

      @Test
        fun  testTopicLessonsFragment_saveInstanceState_verifyCorrectStateRestored() {
          launch<TopicActivity>(
            createTopicPlayStoryActivityIntent(
              internalProfileId,
              TEST_CLASSROOM_ID_1,
              RATIOS_TOPIC_ID,
              RATIOS_STORY_ID_0
            )
          ).use { scenario ->
            clickLessonTab()
            testCoroutineDispatchers.runCurrent()
      
            scrollToPosition(position = 2)
            clickStoryItem(position = 2, targetViewId = R.id.chapter_list_drop_down_icon)
      
            scenario.recreate()
      
            scrollToPosition(position = 2)
            onView(
              atPositionOnView(
                recyclerViewId = R.id.story_summary_recycler_view,
                position = 2,
                targetViewId = R.id.chapter_recycler_view
              )
            ).check(matches(isDisplayed()))
          }
        }

BenHenning replied:
I think this approach of verifying looks really reasonable!

BenHenning committed:
Fix #5473: Add wiki guide for upgrading the app's compile/target SDK versions (#5538)

## Explanation

Fixes #5473

This PR introduces a new wiki page that describes the process by which
we can carefully upgrade the app to a newer compile/target SDK version
(with a specific emphasis on the target SDK version).

This is largely meant to capture past processes we have followed since
those weren't documented in an easily reproducible way. The intent is
for this now-documented process to be trialed by upgrading the app to
SDK 34 in order to address #5535 (though that technically means that
part (1) of the documented process has already been completed).

This PR does not:
- Address exactly when the compile version should be updated, except
when required by the target SDK version. It's expected that the team
will perform an out-of-band analysis for cases when the compile SDK
actually needs to be upgraded sooner than the target SDK.
- Specify a strongly enforced cadence for upgrading the target SDK since
the Google Play Console and Google Play app policy seems to already
perform this function, though it does add a recommendation for checking
for updates periodically (July/August every year).

## Essential Checklist
- [x] The PR title and explanation each start with "Fix #bugnum: " (If
this PR fixes part of an issue, prefix the title with "Fix part of
#bugnum: ...".)
- [x] Any changes to
[scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets)
files have their rationale included in the PR explanation.
- [x] The PR follows the [style
guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android
Studio
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and
is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
N/A -- This is introducing a new wiki page and doesn't change the app
itself.

translatewiki committed:
Localisation updates from https://translatewiki.net. (#5274)

Translation updates

dattasneha committed:
Fix #5329: Add color formatting to static check messages (#5540)

<!-- READ ME FIRST: Please fill in the explanation section below and
check off every point from the Essential Checklist! -->
## Explanation
Fixes #5329  

This PR implements color-coded messages for static checks to improve
experience. Error messages are displayed in red for failing checks,
while warnings are shown in yellow. Successful checks are indicated with
green messages.
<!--
- Explain what your PR does. If this PR fixes an existing bug, please
include
- "Fixes #bugnum:" in the explanation so that GitHub can auto-close the
issue
  - when this PR is merged.
  -->

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If
this PR fixes part of an issue, prefix the title with "Fix part of
#bugnum: ...".)
- [x] Any changes to
[scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets)
files have their rationale included in the PR explanation.
- [x] The PR follows the [style
guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android
Studio
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and
is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related
changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of
the before & after UI changes
- For the screenshots above, include both English and pseudo-localized
(RTL) screenshots (see [RTL
guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see
[accessibility
guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and
dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests
locally & that they're passing

---------

Co-authored-by: Adhiambo Peres <59600948+adhiamboperes@users.noreply.github.com>

TanishMoral11 committed:
Fix #2652 : Improve splash screen implementation with new drawable and updated styles (#5559)

Fixes #2652: This PR aims to enhance the splash screen implementation by
making the following changes:

- **splash_image.xml**: Created a new layer-list drawable that includes
a background color (`@color/oppiaPrimaryLight`) and centers the splash
image (`@drawable/splash_page`).
- **splash_page.xml**: Modified the path data for the splash image and
removed the old clip-path to ensure proper rendering.
- **styles.xml**: Updated the `SplashScreenTheme` to reference the new
`splash_image` drawable and set `adjustViewBounds` to true, allowing for
better handling of different screen sizes.

These changes are intended to improve the visual consistency and
responsiveness of the splash screen across various devices.

## Essential Checklist
- [x] The PR title and explanation each start with "Fix #bugnum: " (If
this PR fixes part of an issue, prefix the title with "Fix part of
#bugnum: ...").
- [x] Any changes to `scripts/assets` files have their rationale
included in the PR explanation.
- [x] The PR follows the [style
guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android
Studio
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and
is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

Before : 

https://github.com/user-attachments/assets/161eb1ca-57b1-4b63-829a-de41f1b99702

After : 

https://github.com/user-attachments/assets/fcd480ff-aabd-4fe3-9887-b9bf50b276ac

---------

Co-authored-by: Adhiambo Peres <59600948+adhiamboperes@users.noreply.github.com>

jainv4156 committed:
Fixes #5484: Multiple substitutions specified in non-positional format of st… (#5560)

## Explanation
Fixes : #5484 

Some string resources have multiple %s . To fix the error, we must
identify each string position with a number : %1$s

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If
this PR fixes part of an issue, prefix the title with "Fix part of
#bugnum: ...".)
- [x] Any changes to
[scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets)
files have their rationale included in the PR explanation.
- [x] The PR follows the [style
guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android
Studio
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and
is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

---------

Co-authored-by: Adhiambo Peres <59600948+adhiamboperes@users.noreply.github.com>

BenHenning commented:
Thanks @subhajitxyz. I followed up to your comments--PTAL.

XichengSpencer committed:
Fix #4419: Remove TODO Marked Files And Other Alpha Kenya Related Files (#5417)

<!-- READ ME FIRST: Please fill in the explanation section below and
check off every point from the Essential Checklist! -->
## Explanation
Fix #4419 
Removed application\alphakenya package as Adhiambo suggests.
Removed codes in bazel files,test_file_exemptions.textproto that is
related with alphakenya
Remove AlphaKenya related modules under android/util/logging including
KenyaAlphaPlatformParameterModule, EventLoggingModule etc. and
corresponding tests and Bazel configs
Inlined StandardEventTypeToHumanReadableNameConverterImpl.kt into
EventTypeToHumanReadableNameConverter.kt
Removed EventLoggingConfigurationModule
<!--
- Explain what your PR does. If this PR fixes an existing bug, please
include
- "Fixes #bugnum:" in the explanation so that GitHub can auto-close the
issue
  - when this PR is merged.
  -->

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If
this PR fixes part of an issue, prefix the title with "Fix part of
#bugnum: ...".)
- [x] Any changes to
[scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets)
files have their rationale included in the PR explanation.
- [x] The PR follows the [style
guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android
Studio
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and
is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related
changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of
the before & after UI changes
- For the screenshots above, include both English and pseudo-localized
(RTL) screenshots (see [RTL
guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see
[accessibility
guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and
dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests
locally & that they're passing

---------

Co-authored-by: Adhiambo Peres <59600948+adhiamboperes@users.noreply.github.com>
Co-authored-by: Ben Henning <ben@oppia.org>

translatewiki committed:
Localisation updates from https://translatewiki.net. (#5569)

Translation updates

manas-yu committed:
Fix #5232: ConsoleLogger overwrites local log file for each line write (#5550)

## Explanation
Fix #5232
This PR updates the `ConsoleLogger` class to improve logging
functionality. The key changes include:

- **Append Mode for Logs**: The `FileWriter` is now opened in append
mode, ensuring that new log entries are added to the existing log file
instead of overwriting it.
- **Long-lived PrintWriter**: A long-lived `PrintWriter` instance has
been implemented to enhance performance by reducing the overhead of
opening and closing the log file multiple times.
- **Close Log File**: A `closeLogFile()` method has been added to
properly close the `PrintWriter`, preventing resource leaks after
logging is complete.

These changes address the issue of logs being overwritten and enhance
performance while ensuring proper resource management.



## Essential Checklist
- [x] The PR title and explanation each start with "Fix #bugnum: " (If
this PR fixes part of an issue, prefix the title with "Fix part of
#bugnum: ...".)
- [x] Any changes to
[scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets)
files have their rationale included in the PR explanation.
- [x] The PR follows the [style
guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android
Studio
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and
is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

---------

Co-authored-by: Adhiambo Peres <59600948+adhiamboperes@users.noreply.github.com>

manas-yu committed:
Fix #5566: Handle action_required status in code coverage (#5574)

<!-- READ ME FIRST: Please fill in the explanation section below and
check off every point from the Essential Checklist! -->
## Explanation
<!--
- Explain what your PR does. If this PR fixes an existing bug, please
include
- "Fixes #bugnum:" in the explanation so that GitHub can auto-close the
issue
  - when this PR is merged.
  -->

Fixes #5566: This PR addresses the issue of code coverage comments not
being posted on pull requests when certain review status ("Request
Changes") result in an `action_required` status.

### Changes:
- Updated `comment_coverage.yml` to include `action_required` in the
`allowed-conclusions` list for the `check_code_coverage_completed` job.
This ensures that the workflow continues even if a review requires
further action.
## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If
this PR fixes part of an issue, prefix the title with "Fix part of
#bugnum: ...".)
- [x] Any changes to
[scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets)
files have their rationale included in the PR explanation.
- [x] The PR follows the [style
guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android
Studio
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and
is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

theayushyadav11 committed:
Fixes #3641 Use finish() instead of intent for smoother navigation (#5557)

## Description:
- Fixes #3641  
This PR simplifies the navigation logic in `AddProfileActivity` and
documents the reasoning behind not using `finish()` in
`ExitProfileDialogFragment`.

### Changes Made:
- **AddProfileActivity:**  
- Replaced the use of `Intent` with `FLAG_ACTIVITY_CLEAR_TOP` by calling
`finish()`.
- With the live data bug resolved, the activity stack can now be managed
more cleanly without requiring a new intent.

  **Benefits:**  
  - Avoids unnecessary activity recreation.  
  - Improves navigation efficiency and reduces overhead.  

- **ExitProfileDialogFragment:**  
- Retained the use of an `Intent` to navigate to
`ProfileChooserActivity` instead of switching to `finish()`.
  - **Reason:**  
- `ProfileChooserActivity` is not directly below the current activity in
the stack.
- If `finish()` were used, the app would incorrectly navigate to
`PinPasswordActivity`, resulting in undesirable behavior.

### Testing:
1. Verified that `AddProfileActivity` correctly navigates back to
`ProfileChooserActivity` using `finish()` without activity recreation.
2. Confirmed that `ExitProfileDialogFragment` behaves as expected,
retaining the correct navigation path through `Intent` to
`ProfileChooserActivity`.

### Summary:
This PR ensures cleaner navigation for `AddProfileActivity` by using
`finish()`, while maintaining correct behavior in
`ExitProfileDialogFragment` by continuing to use an intent-based
approach. This makes the app more efficient while preserving user
experience.

### Video Demo as everything is working fine.

https://github.com/user-attachments/assets/fb383845-5672-4236-8618-a87583cbbba0



## Essential Checklist
- [x] The PR title and explanation each start with "Fix #bugnum: " (If
this PR fixes part of an issue, prefix the title with "Fix part of
#bugnum: ...".)
- [x] Any changes to
[scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets)
files have their rationale included in the PR explanation.
- [x] The PR follows the [style
guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android
Studio
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and
is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

whyash8 committed:
Fix#3146 : Create a generic utility for filtering enums  (#5529)

## Explanation

Fixes: #3146

This PR introduces a new utility function filterByEnumCondition to
standardize filtering of enums across various parts of the oppia-android
codebase. This utility function allows filtering of collections based on
a condition applied to enum values.

This PR introduces a new utility function filterByEnumCondition to
standardize filtering of enums across various parts of the oppia-android
codebase. This utility function allows filtering of collections based on
a condition applied to enum values.

**Key Changes**:
Added Utility Function:
filterByEnumCondition: A generic function to filter a collection based
on a condition applied to an enum extracted from each item in the
collection.

**Updated Existing Code**:
Refactored code in getLeastRecentMetricLogIndex,
getLeastRecentMediumPriorityEventIndex, and other methods to utilize the
new filterByEnumCondition function.
Updated the calculation of completedChapterCount and
inProgressChapterCount using the new utility function.

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [ x] The PR title and explanation each start with "Fix #bugnum: " (If
this PR fixes part of an issue, prefix the title with "Fix part of
#bugnum: ...".)
- [x ] Any changes to
[scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets)
files have their rationale included in the PR explanation.
- [x ] The PR follows the [style
guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x ] The PR does not contain any unnecessary code changes from Android
Studio
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [ x] The PR is made from a branch that's **not** called "develop" and
is up-to-date with "develop".
- [x ] The PR is **assigned** to the appropriate reviewers
([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).
-

---------

Co-authored-by: Adhiambo Peres <59600948+adhiamboperes@users.noreply.github.com>

subhajitxyz committed:
do correction as per review

subhajitxyz committed:
fix

subhajitxyz committed:
fix

adhiamboperes commented:
@subhajitxyz, do we need to reopen this PR?

subhajitxyz commented:
I have created a new PR with all the requested review changes. Therefore, we don't need to reopen this PR.
